proto/v1beta1/intoto.proto

```diff
@@ -0,0 +1,145 @@
+// Copyright 2020 The Grafeas Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package grafeas.v1beta1.intoto;
+
+option go_package = "github.com/grafeas/grafeas/proto/v1beta1/intoto_go_proto";
+option java_multiple_files = true;
+option java_package = "io.grafeas.v1beta1.intoto";
+option objc_class_prefix = "GRA";
+
+import "google/protobuf/struct.proto";
+
+// This file defines the proto that matches the schema of an in-toto link
+// file. in-toto is a framework to protect a software supply chain. You can read
+// more about in-toto at https://in-toto.io/ and you can find the current
+// version of the in-toto specification at
+// https://github.com/in-toto/docs/blob/master/in-toto-spec.md
+
+
+// A signature object consists of the KeyID used and the signature itself.
+message Signature {
+  string keyid = 1 [json_name="keyid"];
+  string sig = 2 [json_name="sig"];
+}
+
+// An object defined to capture the by products of a step.
+message ByProductObject {
+  int32 return_value = 1 [json_name="return-value"];
+  string stderr = 2 [json_name="stderr"];
+  string stdout = 3 [json_name="stdout"];
+}
+
+// This corrresponds to the core in-toto link attestation.
+message Link {
+  // This field contains the full command executed for the step. This can also
+  // be empty if attestations are generated for operations that aren't directly
+  // mapped to a specific command. Each term in the command is an independent
+  // string in the list. An example of a command in the in-toto metadata field
+  // is:
+  // "command": ["git", "clone", "https://github.com/in-toto/demo-project.git"]
+  repeated string command = 1 [json_name="command"];
+
+  // Materials are the supply chain artifacts that go into the step and are used
+  // for the operation performed. The key of the map is the path of the artifact
+  // and the structure contains the recorded hash information. An example is:
```

Comment:
re "structure contains the recorded hash information": would it make sense to define it here as well then? If it's only ever going to be a hash, we have ways to represent this.

Reply:
We don't have a strict structure for it as we don't enforce specific hash algorithms. It can also be multiple as an artifact can have its hashes recorded using more than one algorithm.

Reply:
Also, changes here **_could_** mean that when it's serialised into a JSON from the Grafeas server, it would no longer match the in-toto format and signature verification will fail.

Reply:
Makes sense, thanks for clarifying!